### PR TITLE
Issue - 996 | Refactor BasicAuthMiddleware to reduce cognitive complexity

### PR DESCRIPTION
**Pull request by nidhey27**
## Issue
Closes: #996 

### **Description:**

- Extracted `parseBasicAuth` to handle authorization header parsing
- Added `respondUnauthorized` for consistent unauthorized responses
- Moved credential validation logic to `validateCredentials`
- Simplified main middleware function for improved readability and maintainability

**Checklist:**

-   [x] I have formatted my code using  `goimport`  and  `golangci-lint`.
-   [x] All new code is covered by unit tests.
-   [x] This PR does not decrease the overall code coverage.
-   [x] I have reviewed the code comments and documentation for clarity.

### CODE REVIEW

**Comment by Umang01-hash** [comment]
@nidhey27 your PR has some failing tests as well as code quality issues. Please resolve them

**Comment by nidhey27** [comment]
@Umang01-hash sure I will have a look at it.

File: pkg/gofr/http/middleware/basic_auth_test.go [hunk 1]
```diff
@@ -179,3 +180,144 @@ func Test_BasicAuthMiddleware_well_known(t *testing.T) {
...
+	handler := http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	})
+
+	req := httptest.NewRequest(http.MethodGet, "/", http.NoBody)
+	rr := httptest.NewRecorder()
+	middleware(handler).ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusUnauthorized, rr.Code)
+	assert.Contains(t, rr.Body.String(), "Unauthorized: Authorization header missing")
```
Comment (vipul-rawat) [inline pkg/gofr/http/middleware/basic_auth_test.go:322]: This test condition is failing as, on missing header the checks respond with - "Invalid or missing header"

**Comment by nidhey27** [comment]
@Umang01-hash @vipul-rawat Were the recent changes you made intentional or accidental?

**Comment by Umang01-hash** [comment]
@nidhey27 We haven't committed anything intentionally. I think you pulled from development and all our recent commits there got merged.

File: pkg/gofr/http/middleware/basic_auth.go [hunk 2]
```diff
@@ -28,56 +28,61 @@ func BasicAuthMiddleware(basicAuthProvider BasicAuthProvider) func(handler http.
...
+	if !found { // Ensure both username and password are returned as empty if colon separator is missing
+		return "", "", false
 	}
+
+	return username, password, true
+}
+
+// respondUnauthorized sends a 401 Unauthorized response with a given message.
+func respondUnauthorized(w http.ResponseWriter, message string) {
+	http.Error(w, "Unauthorized: "+message, http.StatusUnauthorized)
```
Comment (vikash) [inline pkg/gofr/http/middleware/basic_auth.go:75]: function with single line in its body may not be required.

**Comment by Umang01-hash** [comment]
@nidhey27 Please resolve the review comment given so that we can proceed ahead and merge your PR.

**Comment by Umang01-hash** [comment]
Screenshot 2024-11-29 at 1 25 53 PM

@nidhey27 You still have one code quality issue in the PR. Please resolve this. Rest everything looks good.

Thankyou

**Review by Umang01-hash — APPROVED** [review]

**Pull request merged**